File: database-driver-plugin/src/main/java/de/lino/database/database/file/DefaultFileProvider.java
package de.lino.database.database.file;

import de.lino.database.json.file.FileProvider;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

/**
 * The concrete, NIO-backed {@link FileProvider}: every operation is implemented directly on top
 * of {@link Files} (or, for bulk copies, {@link Files#walkFileTree}), rather than the legacy
 * {@code java.io} file API. Installs itself as {@link FileProvider}'s generated
 * {@code getInstance()} accessor on construction.
 * <p>
 * Stateless beyond the inherited singleton instance, so every method is safe to call
 * concurrently from multiple threads; thread-safety of the operations themselves is delegated
 * entirely to the filesystem.
 */
public class DefaultFileProvider extends FileProvider {

    /**
     * Installs this instance as {@link FileProvider}'s generated {@code getInstance()} accessor.
     */
    public DefaultFileProvider() {
        setInstance(this);
    }

    @Override
    public void deleteFile(@NotNull final File file) {
        deleteFile(file.toPath());
    }

    @Override
    public void deleteFile(@NotNull final Path file) {

        try {
            Files.deleteIfExists(file);
        } catch (final IOException exception) {
            exception.printStackTrace();
        }

    }

    @Override
    public void createFile(@NotNull final Path path) {

        if (Files.exists(path)) return;

        try {

            final Path parent = path.getParent();
            if (parent != null) Files.createDirectories(parent);

            Files.createFile(path);

        } catch (final IOException exception) {
            exception.printStackTrace();
        }

    }

    @Override
    public void createFile(@NotNull final File file) {
        createFile(file.toPath());
    }

    @Override
    public void updateFile(@NotNull final Path file) {
        deleteFile(file);
        createFile(file);
    }

    @Override
    public void updateFile(@NotNull final File file) {
        deleteFile(file);
        createFile(file);
    }

    @Override
    public void rename(@NotNull final File file, @NotNull final String newName) {

        try {
            Files.move(file.toPath(), Path.of(newName));
        } catch (final IOException exception) {
            exception.printStackTrace();
        }

    }

    @Override
    public void createDirectory(@Nullable final Path path) {

        if (path == null) return;

        try {
            Files.createDirectories(path);
        } catch (final IOException exception) {
            exception.printStackTrace();
        }

    }

    @Override
    public void doCopy(@NotNull final String from, @NotNull final String target) {

        try {
            Files.copy(Path.of(from), Path.of(target), StandardCopyOption.REPLACE_EXISTING);
        } catch (final IOException exception) {
            exception.printStackTrace();
        }

    }

    @Override
    public void deleteAllFilesInDirectory(@NotNull final Path dirPath) {

        List<Path> files = List.of();

        try (final Stream<Path> entries = Files.list(dirPath)) {
            files = entries.filter(Files::isRegularFile).toList();
        } catch (final IOException exception) {
            exception.printStackTrace();
        }

        // Listed and closed above before deleting, rather than deleting while the directory
        // stream is still open - mutating a directory while iterating a live handle on it is
        // platform-dependent behavior under NIO.
        files.forEach(this::deleteFile);

    }

    @Override
    public void deleteDirectory(@NotNull final Path dirPath) {

        List<Path> children = List.of();

        try (final Stream<Path> entries = Files.list(dirPath)) {
            children = entries.toList();
        } catch (final IOException exception) {
            exception.printStackTrace();
        }

        for (final Path path : children) {
            if (Files.isDirectory(path)) {
                deleteDirectory(path);
            } else {
                deleteFile(path);
            }
        }

        deleteFile(dirPath);

    }

    @Override
    public void recreateDirectory(@NotNull final Path path) {

        if (Files.exists(path)) {
            if (path.toFile().isDirectory()) {
                deleteDirectory(path);
            } else {
                deleteFile(path.toFile());
            }
        }

        createDirectory(path);

    }

    @Override
    public void copyDirectory(@NotNull final Path path, @NotNull final Path target) {
        copyDirectory(path, target, Set.of());
    }

    @Override
    public void copyDirectory(@NotNull final Path path, @NotNull final Path target, @NotNull final Collection<String> excludedFiles) {

        final Set<String> excluded = Set.copyOf(excludedFiles);

        try {

            Files.walkFileTree(path, new SimpleFileVisitor<>() {

                @Override
                public FileVisitResult visitFile(final Path file, final BasicFileAttributes attrs) throws IOException {

                    if (excluded.contains(file.getFileName().toString())) return FileVisitResult.CONTINUE;

                    final Path targetFile = target.resolve(path.relativize(file));
                    final Path parent = targetFile.getParent();

                    if (parent != null) Files.createDirectories(parent);

                    Files.copy(file, targetFile, StandardCopyOption.REPLACE_EXISTING);
                    return FileVisitResult.CONTINUE;

                }
            });

        } catch (final IOException exception) {
            exception.printStackTrace();
        }

    }

}

File: python/database-driver-plugin/src/database_driver/plugin/database/file/default_file_provider.py
"""Mirror of ``de.lino.database.database.file.DefaultFileProvider``."""

from __future__ import annotations

import shutil
import traceback
from collections.abc import Collection
from pathlib import Path

from database_driver.api.json.file.file_provider import FileProvider, PathLike


class DefaultFileProvider(FileProvider):
    """The concrete, ``pathlib``-backed ``FileProvider``: every operation is implemented
    directly on top of :mod:`pathlib`/:mod:`shutil` - the Python counterpart of the Java
    edition's NIO implementation. Installs itself as ``FileProvider``'s singleton
    accessor on construction.

    Stateless beyond the inherited singleton instance, so every method is safe to call
    concurrently from multiple threads; thread-safety of the operations themselves is
    delegated entirely to the filesystem. Failures are printed rather than raised,
    matching the Java edition's log-and-continue convention throughout.
    """

    def __init__(self) -> None:
        FileProvider.set_instance(self)

    def delete_file(self, file: PathLike) -> None:
        try:
            Path(file).unlink(missing_ok=True)
        except OSError:
            traceback.print_exc()

    def create_file(self, file: PathLike) -> None:
        path = Path(file)
        if path.exists():
            return
        try:
            if path.parent != Path():
                path.parent.mkdir(parents=True, exist_ok=True)
            path.touch()
        except OSError:
            traceback.print_exc()

    def update_file(self, file: PathLike) -> None:
        self.delete_file(file)
        self.create_file(file)

    def rename(self, file: PathLike, new_name: str) -> None:
        try:
            Path(file).rename(new_name)
        except OSError:
            traceback.print_exc()

    def create_directory(self, path: PathLike | None) -> None:
        if path is None:
            return
        try:
            Path(path).mkdir(parents=True, exist_ok=True)
        except OSError:
            traceback.print_exc()

    def do_copy(self, from_path: str, target: str) -> None:
        try:
            shutil.copyfile(from_path, target)
        except OSError:
            traceback.print_exc()

    def delete_all_files_in_directory(self, dir_path: PathLike) -> None:
        # Listed fully before deleting, rather than deleting while iterating a live
        # directory handle - the same ordering the Java edition documents for NIO.
        files: list[Path] = []
        try:
            files = [entry for entry in Path(dir_path).iterdir() if entry.is_file()]
        except OSError:
            traceback.print_exc()
        for file in files:
            self.delete_file(file)

    def delete_directory(self, dir_path: PathLike) -> None:
        try:
            shutil.rmtree(dir_path, ignore_errors=True)
        except OSError:
            traceback.print_exc()

    def recreate_directory(self, path: PathLike) -> None:
        target = Path(path)
        if target.exists():
            if target.is_dir():
                self.delete_directory(target)
            else:
                self.delete_file(target)
        self.create_directory(target)

    def copy_directory(self, path: PathLike, target: PathLike, excluded_files: Collection[str] = ()) -> None:
        excluded = set(excluded_files)
        source = Path(path)
        destination = Path(target)
        try:
            for file in source.rglob("*"):
                if not file.is_file() or file.name in excluded:
                    continue
                target_file = destination / file.relative_to(source)
                target_file.parent.mkdir(parents=True, exist_ok=True)
                shutil.copyfile(file, target_file)
        except OSError:
            traceback.print_exc()
